Release/extensibility

soulshined opened the pull request:
This is a lot to take in - but basically instead of creating my own package I thought I would just not reinvent the wheel and add on to yours. So thanks for the start on this and sharing! I would recommend looking through the commits individually to make sense of them but if you have questions, feel free to reject and comment your concerns

TLDR:

- Add's caching support, files should not be read every single time a static method is called, especially when you are doing multiple validations
- Add support for validation against an array of values
- Add support for multi-validation (meaning, validate a provided string is either banned in words or usernames for examples)
- Added unit tests 
- Allow a way to extend banned data. Hard coded things (only) in 2022 just makes it less maintainable imo

I don't know what your time looks like, but it would be good to consider and discuss some other things, like stripping acronyms, because a user could simply type: `B.I.T.C.H` or simply with spaces : `B I TCH` 

It might be hard, but we should also consider locale based strings, because a user could easily bypass a banned word using accents, for example. Like `bítch`

I do not address those in this PR, but it is something to consider for future state if you plan on keeping this open source, otherwise i may just fork this and put it on packagist and maintain it with sqlite3 or something portable

---

## Commits 

- [`ed57e33`] **Update README examples for brevity**

- [`200e633`] **Add support to validate that all provided values in a set are banned**
  - Add a `isAny(array $value, bool $email = false, bool $word = false, bool $username = false, bool $ip = false, bool $bank_accounts = false)` method
      - This method returns true if any of the provided values is banned in any of the paths selected
      - This method short circuits
  
  - Add a `isAll(array $value, bool $email = false, bool $word = false, bool $username = false, bool $ip = false, bool $bank_accounts = false)` method
      - This method returns true if ALL of the provided values are banned across all the paths selected
      - This method short circuits
  - Add unit tests accordingly

- [`79bdd34`] **Update isEmail logic to ensure the flag does not persist across non-email validations**
  - It's entirely possible that a username has special characters, or @ symbol
  - Ensured the isEmail flag is set to false post validation

- [`c04470c`] **Add support for extending banned content**
  - Added a way to merge value[s] and user defined/provided files
  - Added unit tests accordingly

- [`cf3e1f4`] **Add support for multiple value validation**
  - Changed each public method signature to allow intake of string or array of values
  - Changed the entire Ban control flow to be more functional programming based - to support
  multiple value validation
  - Add unit tests for intake of array of values accordingly

- [`1488547`] **Add caching control flow**
  - Files should not be read every request, especially when the methods only
  allow searching for singular matches at a given time. Subsequent requests would read the same file
  - Added static variable called $cache where the filenames serve as the key to query in readFile()

soulshined committed:
Add caching control flow

- Files should not be read every request, especially when the methods only
allow searching for singular matches at a given time. Subsequent requests would read the same file
- Added static variable called $cache where the filenames serve as the key to query in readFile()

soulshined committed:
Add support for multiple value validation

- Changed each public method signature to allow intake of string or array of values
- Changed the entire Ban control flow to be more functional programming based to support
multiple value validation
- Add unit tests for intake of array of values accordingly

soulshined committed:
Add support for extending banned content

- Added a way to merge value[s] and user defined/provided files
- Added unit tests accordingly

soulshined committed:
Update isEmail logic to ensure the flag does not persist across non-email validations

- It's entirely possible that a username has special characters, or @ symbol
- Ensured the isEmail flag is set to false post validation

soulshined committed:
Add support to validate that all provided values in a set are banned

- Add a `isAny(array $value, bool $email = false, bool $word = false, bool $username = false, bool $ip = false, bool $bank_accounts = false)` method
    - This method returns true if any of the provided values is banned in any of the paths selected
    - This method short circuits

- Add a `isAll(array $value, bool $email = false, bool $word = false, bool $username = false, bool $ip = false, bool $bank_accounts = false)` method
    - This method returns true if ALL of the provided values are banned across all the paths selected
    - This method short circuits
- Add unit tests accordingly

soulshined committed:
Update README examples for brevity

pH-7 commented on src/Ban.php:
```diff
@@ -18,105 +20,211 @@ class Ban
     private const BANK_ACCOUNT_FILE = 'bank_accounts.txt';
     private const IP_FILE = 'ips.txt';
     private const COMMENT_SIGN = '#';
+    private static array $cache = [
```
Typed properties requires PHP 7.4+, but I'm happy with this 🙂 Let's update the required PHP version in [composer.json](https://github.com/pH-7/NotAllowed/blob/main/composer.json#L27) (as well as in the README), and we will be good to go.

soulshined replied:
I'm completely fine with leaving everything PHP 7 compatible. I was just in auto pilot mode when typing and I personally use 8 so it was second nature.

pH-7 replied:
Sure, no worries! I will move this library to support PHP 8 or higher and drop PHP 7+4 for the next release, so that's all good. Thank you

pH-7 commented on src/Ban.php:
```diff
@@ -18,105 +20,211 @@ class Ban
+     *
+     * @return bool true if the value, or any of array values, are banned based on chosen validation paths
+     */
+    public static function isAny(string | array $value,
```
With the union types, we will actually need at least PHP 8 for this new version of the package. Good additions 👍

pH-7 replied:
I also see your comment mentioning PHP 8 concerning the named arguments. Thanks for commenting when needed!

pH-7 commented on src/Ban.php:
```diff
@@ -18,105 +20,211 @@ class Ban
+                                 bool $ip = false,
+                                 bool $bank_accounts = false) : bool {
+
+        if ($email && static::isEmail($value)) return true;
```
Let's use curly brackets here on if-statements, so we comply with PSR-2 and keep our code consistent 🙂

pH-7 commented:
Hi @soulshined, thanks very much for opening a PR here!! Your additions look fantastic!! 🎉 I will have a closer look and review it later this week and I will leave my comments if so by Friday. Enjoy your holidays.


Thanks again for contributing to the library!

pH-7 commented on tests/BanTest.php:
```diff
@@ -74,9 +118,80 @@ public function testIsBannedIp(string $sIp): void
+        $this->assertTrue(Ban::isAny(['good', 'good2', 'bitch', 'good3', 'a@tafmail.COM'], true, true));
+    }
+
+    public function testIsAll() {
```
No biggie here, but we could have more specific test names here. `testIsAny`, `testIsAll` is quite vague and could lead to confusion when looking at test names over the terminal.

soulshined replied:
I'll leave this up to you for recommendations then. I'm indifferent and don't really have a preference to test names when I code. I just tried to keep it similar to what you had for other test names. If you want to throw some out there I can update, or you can just commit to this pr directly if that's what you prefer

pH-7 commented on src/Ban.php:
```diff
@@ -18,105 +20,211 @@ class Ban
 
-    /** @var string */
-    private static $sFile;
+    /** @var bool */
```
Again, no biggie. Just that this inline comment could be removed now as it’s redundant with the typed property 🙂

pH-7 commented on tests/extended_usernames.txt:
```diff
@@ -0,0 +1,4 @@
+rickastley1987
```
To make things tidier in `/tests`, we could move these two files in a `dataset` folder

pH-7 commented on tests/BanTest.php:
```diff
@@ -74,9 +118,80 @@ public function testIsBannedIp(string $sIp): void
+    }
+
+    public function testExtendedFileIsMerged() {
+        Ban::mergeFile('usernames', __DIR__ . '/./extended_usernames.txt');
```
Was there a reason for this? I don’t think we would need `/./`, just `__DIR__ . '/extended_usernames.txt'` 🙂

soulshined replied:
It's just a habit. Not an issue to remove

pH-7 commented on src/Ban.php:
```diff
@@ -18,105 +20,211 @@ class Ban
+        return static::$cache[$scope];
+    }
+
+    private static function readFile(string $path) {
```
Perhaps, we could still have a return type here… `readFile(string $path): array`

soulshined committed:
Remove bIsEmail flag

soulshined committed:
Remove property types and union types

soulshined committed:
Add curly braces

soulshined committed:
Add return type to readFile

soulshined committed:
Moved test extended files to banned-data subdir

pH-7 commented:
Happy New Year @soulshined! 🎉 Beautiful! I will merge it for now @soulshined and make some minor changes if needed. Looks really great! Thank you so much for contributing to this library. Any feedback or further enhancements are more than welcome!

pH-7 commented:
<img width="831" alt="Screen Shot 2023-01-01 at 8 12 17 pm" src="https://user-images.githubusercontent.com/1325411/210165950-cff87ff9-497c-4b80-b104-8df1e943155f.png">

EDIT: Unit tests are now passing. Just dropped the `__DIR__`

pH-7 committed:
Fixing path

We would need to mention the `/tests` folder as the tests are being ran from the main project folder

pH-7 committed:
Fixing path `Ban::mergeFile` to avoid `realpath` returning FALSE

Fixing the path to avoid `realpath` giving FALSE when running unit tests

soulshined commented:
> Happy New Year @soulshined! 🎉 Beautiful! I will merge it for now @soulshined and make some minor changes if needed. Looks really great! Thank you so much for contributing to this library. Any feedback or further enhancements are more than welcome!

Thanks. Happy New Year to as well! 

I may have misunderstood though unfortunately. Based on the initial feedback I went ahead and reverted parameter hints (like the union types) in trying to be backwards compatible with 7.

I think I should add those back now that minimum version is 8

Also since we're only supporting 8, I think another enhancement could be to use an enum for the 'cache' keys. The secondary benefit there is users would know an enum is used for 'scope' parameters and wouldn't have to stumple upon a doc comment to know what possible values are allowed 

Lastly. Can I recommend to you to start adding versions to the composer json file so that consumers aren't coerced into using new changes? 

If these sound good to you I can do a final PR on that, if not no worries

pH-7 commented:
Hi @soulshined! 👋 Hope you are well.

Sorry for my late reply. Got a quite busy week! 

> I may have misunderstood though unfortunately. Based on the initial feedback I went ahead and reverted parameter hints (like the union types) in trying to be backwards compatible with 7.

Sorry for this misunderstanding. I believe it's a better idea to drop the compatibility with PHP 7 and move forward to v8 or v8.1. PHP 8 has been released for over two years now.

> I think I should add those back now that minimum version is 8

Yes, that's right 👍 

> Also since we're only supporting 8, I think another enhancement could be to use an enum for the 'cache' keys. The secondary benefit there is users would know an enum is used for 'scope' parameters and wouldn't have to stumple upon a doc comment to know what possible values are allowed

That's an excellent idea indeed! I've made some minor changes (you will see on the `main` branch), where I've stored them in private constants. However, having an enum instead is a cleaner and more appropriate approach here! I'm happy 

> Lastly. Can I recommend to you to start adding versions to the composer json file so that consumers aren't coerced into using new changes?

I was planning to create a `v2.0.0` tag once everything is completed. I don't think we also need to add a `version` prop in `composer.json` Did you mean this https://getcomposer.org/doc/04-schema.md#version ?

> If these sound good to you I can do a final PR on that, if not no worries

I'm very happy with your suggestions. Feel free to go ahead. You will just need to rebase the `main` branch as I've made some minor changes in there.


Have a great weekend ahead.

pH-7 commented:
@soulshined Just so you know, I have started implementing an enum (called `BannedType` for now). In case you would like to enhance it, you will need to rebase your branch with `main` first.

Let me know if needed 🙂
